### PR TITLE
feat(abac): Add support for ABAC Conditions

### PR DESCRIPTION
**Pull request by jon-whit**
## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: pkg/go/transformer/dsltojson.go [hunk 1]
```diff
@@ -59,6 +64,61 @@ func (l *OpenFgaDslListener) EnterTypeDef(ctx *parser.TypeDefContext) {
...
+
+	l.currentCondition = &pb.Condition{
+		Name:       ctx.ConditionName().GetText(),
+		Expression: "",
+		Parameters: map[string]*pb.ConditionParamTypeRef{},
+	}
+}
+
+func (l *OpenFgaDslListener) ExitConditionParameter(ctx *parser.ConditionParameterContext) {
+	paramContainer := ctx.ParameterType().CONDITION_PARAM_CONTAINER()
```
Comment (miparnisari) [inline pkg/go/transformer/dsltojson.go:84]: This line can panic

```shell
[1/11/23 3:58:38] ~/GitHub/openfga (feat/abac) $ cat demo.fga.yaml 
model: |
  model
    schema 1.1

  type user

  type document
    relations
      define can_view: [user, user with correct_ip]

  condition correct_ip(ip:string) {
     ip == "192.168.0.1"
  }
tuples:
    - user: "user:jon"
      relation: "can_view"
      object: "document:1"
      condition:
        name: correct_ip
tests:
  - check:
      - user: "user:jon"
        object: "document:1"
        context:
          ip: "192.168.0.1"
        assertions:
          can_view: true%                                                                                  [1/11/23 3:58:40] ~/GitHub/openfga (feat/abac) $ ~/GitHub/cli/dist/fga model test --tests demo.fga.yaml 
(PASSING) : Checks (1/1 passing) | ListObjects (0/0 passing)
[1/11/23 3:58:42] ~/GitHub/openfga (feat/abac) $ cat demo.fga.yaml                                      
model: |
  model
    schema 1.1

  type user

  type document
    relations
      define can_view: [user, user with correct_ip]

  condition correct_ip(ip) {
     ip == "192.168.0.1"
  }
tuples:
    - user: "user:jon"
      relation: "can_view"
      object: "document:1"
      condition:
        name: correct_ip
tests:
  - check:
      - user: "user:jon"
        object: "document:1"
        context:
          ip: "192.168.0.1"
        assertions:
          can_view: true%                                                                                  [1/11/23 3:58:47] ~/GitHub/openfga (feat/abac) $ ~/GitHub/cli/dist/fga model test --tests demo.fga.yaml 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x1614e90]

goroutine 1 [running]:
github.com/openfga/language/pkg/go/transformer.(*OpenFgaDslListener).ExitConditionParameter(0xc00039e6c0, 0xc0005dec80)
        /Users/mparnisari/go/pkg/mod/github.com/openfga/language/pkg/go@v0.0.0-20231026200902-6e91eb1b9497/transformer/dsltojson.go:83 +0x110
github.com/openfga/language/pkg/go/gen.(*ConditionParameterContext).ExitRule(0xc00035f5b8?, {0x1d27d28?, 0xc00039e6c0?})
        /Users/mparnisari/go/pkg/mod/github.com/openfga/language/pkg/go@v0.0.0-20231026200902-6e91eb1b9497/gen/openfga_parser.go:6865 +0x3f
github.com/antlr4-go/antlr/v4.(*ParseTreeWalker).ExitRule(0xc0005de980?, {0x1d27d28, 0xc00039e6c0}, {0x4a9af2d0?, 0xc0005dec80?})
        /Users/mparnisari/go/pkg/mod/github.com/antlr4-go/antlr/v4@v4.13.0/tree.go:241 +0x62
github.com/antlr4-go/antlr/v4.(*ParseTreeWalker).Walk(0xc0005de900?, {0x1d27d28, 0xc00039e6c0}, {0x1d2bf58?, 0xc0005dec80})
        /Users/mparnisari/go/pkg/mod/github.com/antlr4-go/antlr/v4@v4.13.0/tree.go:225 +0x179
github.com/antlr4-go/antlr/v4.(*ParseTreeWalker).Walk(0xc00018ae80?, {0x1d27d28, 0xc00039e6c0}, {0x1d2bfa0?, 0xc0005de980})
        /Users/mparnisari/go/pkg/mod/github.com/antlr4-go/antlr/v4@v4.13.0/tree.go:223 +0x125
github.com/antlr4-go/antlr/v4.(*ParseTreeWalker).Walk(0xc00061e3e8?, {0x1d27d28, 0xc00039e6c0}, {0x1d2bfe8?, 0xc0005de900})
        /Users/mparnisari/go/pkg/mod/github.com/antlr4-go/antlr/v4@v4.13.0/tree.go:223 +0x125
github.com/antlr4-go/antlr/v4.(*ParseTreeWalker).Walk(0xc000410358?, {0x1d27d28, 0xc00039e6c0}, {0x1d2c8e8?, 0xc00018ae80})
        /Users/mparnisari/go/pkg/mod/github.com/antlr4-go/antlr/v4@v4.13.0/tree.go:223 +0x125
github.com/openfga/language/pkg/go/transformer.ParseDSL({0xc00014e140?, 0xc00035f850?})
        /Users/mparnisari/go/pkg/mod/github.com/openfga/language/pkg/go@v0.0.0-20231026200902-6e91eb1b9497/transformer/dsltojson.go:376 +0x4c5
github.com/openfga/language/pkg/go/transformer.TransformDSLToJSON({0xc00014e140?, 0x28?})
        /Users/mparnisari/go/pkg/mod/github.com/openfga/language/pkg/go@v0.0.0-20231026200902-6e91eb1b9497/transformer/dsltojson.go:383 +0x18
github.com/openfga/cli/internal/authorizationmodel.(*AuthzModel).ReadFromDSLString(0xc00035f8c0?, {0xc00014e140?, 0x1db?})
        /Users/mparnisari/GitHub/cli/internal/authorizationmodel/model.go:156 +0x2a
github.com/openfga/cli/internal/storetest.getLocalServerAndModel({{0x0, 0x0}, {0xc00014e140, 0xa0}, {0x0, 0x0}, {0xc000052700, 0x1, 0x1}, {0xc000456150, ...}}, ...)
        /Users/mparnisari/GitHub/cli/internal/storetest/localstore.go:103 +0x11b
github.com/openfga/cli/internal/storetest.RunTests(0x0?, {{0x0, 0x0}, {0xc00014e140, 0xa0}, {0x0, 0x0}, {0xc000052700, 0x1, 0x1}, ...}, ...)
        /Users/mparnisari/GitHub/cli/internal/storetest/tests.go:38 +0x5e
github.com/openfga/cli/cmd/model.glob..func3(0xc00042eb00?, {0x1ba60b7?, 0x4?, 0x1ba60bb?})
        /Users/mparnisari/GitHub/cli/cmd/model/test.go:68 +0x566
github.com/spf13/cobra.(*Command).execute(0x25266a0, {0xc00007c6e0, 0x2, 0x2})
        /Users/mparnisari/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0x87c
github.com/spf13/cobra.(*Command).ExecuteC(0x2525560)
        /Users/mparnisari/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3a5
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/mparnisari/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
github.com/openfga/cli/cmd.Execute()
        /Users/mparnisari/GitHub/cli/cmd/root.go:47 +0x1a
main.main()
        /Users/mparnisari/GitHub/cli/main.go:21 +0xf
[1/11/23 3:58:50] ~/GitHub/openfga (feat/abac) $ 

```

**Review by miparnisari — CHANGES_REQUESTED** [review]

**Review by miparnisari — DISMISSED** [review]
If one of the Checks fails, and it had a `context`, can we print that too?

image

File: pkg/go/transformer/dsltojson.go (hunk 1 above)
Comment (rhamzeh) [reply pkg/go/transformer/dsltojson.go:104]: This has been fixed

**Comment by rhamzeh** [comment]
> If one of the Checks fails, and it had a `context`, can we print that too?

@miparnisari can you raise that issue in the CLI?

**Review by jpadilla — APPROVED** [review]

**Pull request merged**